### PR TITLE
Jetpack Section: Restore Status and Restore Complete

### LINKED ISSUE #15191: Jetpack Section: Backup and Restore detail
Project: https://github.com/wordpress-mobile/WordPress-iOS/issues/15193

Screen Shot 2021-01-04 at 13 19 42

#### Network
- [x] Add endpoints (https://github.com/wordpress-mobile/WordPressKit-iOS/pull/321)

#### General
- [x] Replace occurrences of Rewind with Restore to be inline with current branding (#15817)

#### Activity Log
- [x] Update backup table cell to use ⋮ ellipsis icon (https://github.com/wordpress-mobile/WordPress-iOS/pull/15555)
- [x] Tap ⋮ ellipsis icon > present action sheet (https://github.com/wordpress-mobile/WordPress-iOS/pull/15555)
- [x] Tap `Restore` > present Restore Options (https://github.com/wordpress-mobile/WordPress-iOS/pull/15555)
- [x] Tap `Download backup` > present Backup Options (https://github.com/wordpress-mobile/WordPress-iOS/pull/15555)
- [x] Display restore progress cell after returning to activity log from restore flow (https://github.com/wordpress-mobile/WordPress-iOS/pull/15689)
- [x] Display backup progress cell after returning to activity log from backup flow (#15776)

#### Activity Details
- [x] Route `Restore` to Restore Options (https://github.com/wordpress-mobile/WordPress-iOS/pull/15598)
- [x] Add `Download backup` cell (https://github.com/wordpress-mobile/WordPress-iOS/pull/15683)
- [x] Route `Download backup` to Backup Options (https://github.com/wordpress-mobile/WordPress-iOS/pull/15683)

#### Download / Restore Options View
- [x] Create a reusable view controller for both `Restore` and `Download Backup` actions (https://github.com/wordpress-mobile/WordPress-iOS/pull/15566)
- [x] Tap `Cancel` > dismiss (https://github.com/wordpress-mobile/WordPress-iOS/pull/15566)
- [x] Tap `Download` button > push Backup Status  (https://github.com/wordpress-mobile/WordPress-iOS/pull/15566)
- [x] Tap `Restore` button > push Restore Warning (https://github.com/wordpress-mobile/WordPress-iOS/pull/15566)
- [x] ~Check with designer if the entire view should be scrollable or just the download options~ Entire view should be scrollable

#### Restore Warning View
- [x] Tap `Confirm` > send request, push Restore Status (https://github.com/wordpress-mobile/WordPress-iOS/pull/15598)
- [x] Tap `Cancel` > go back to Restore Options (https://github.com/wordpress-mobile/WordPress-iOS/pull/15598)

#### Download / Restore Status View
- [x] Create a new view controller (https://github.com/wordpress-mobile/WordPress-iOS/pull/15598)
- [x] Optimistically display the progress at 0% (https://github.com/wordpress-mobile/WordPress-iOS/pull/15633)
- [x] Add timer to poll the status every x seconds (https://github.com/wordpress-mobile/WordPress-iOS/pull/15633)
- [x] Add progress bar update (https://github.com/wordpress-mobile/WordPress-iOS/pull/15633)
- [x] Error handling
    - [x] No internet connection (https://github.com/wordpress-mobile/WordPress-iOS/pull/15673)
    - [x] Backup / Restore already running (https://github.com/wordpress-mobile/WordPress-iOS/pull/15689)
    - [x] Backup / Restore request failed (https://github.com/wordpress-mobile/WordPress-iOS/pull/15673)
    - [x] Backup / Restore status update failed (https://github.com/wordpress-mobile/WordPress-iOS/pull/15673)
    - [x] Restore failed (https://github.com/wordpress-mobile/WordPress-iOS/pull/15673)
 
#### Download / Restore Success View
- [x] Create a reusable view controller (https://github.com/wordpress-mobile/WordPress-iOS/pull/15633)
- [x] Add customizable button (https://github.com/wordpress-mobile/WordPress-iOS/pull/15633)
- [x] Add button actions
    - [x] Restore: Visit site (https://github.com/wordpress-mobile/WordPress-iOS/pull/15657)
    - [x] Backup: Share link (https://github.com/wordpress-mobile/WordPress-iOS/pull/15657)
    - [x] Backup: Download file / Open link in Safari (https://github.com/wordpress-mobile/WordPress-iOS/pull/15598)

#### Download / Restore Failed View
- [x] Add backup failure state (https://github.com/wordpress-mobile/WordPress-iOS/pull/15673)

#### Notifications Tab
- [x] Override the existing views in notifications tab to deep link to the download view (#15733)

#### Tracks
- [x] Add tracking: https://github.com/wordpress-mobile/WordPress-iOS/pull/15791

#### Finishing Touches
- [x] Check dark mode: https://github.com/wordpress-mobile/WordPress-iOS/pull/15809
- [x] Check Accessibility: https://github.com/wordpress-mobile/WordPress-iOS/pull/15816
- [x] Check iPad: https://github.com/wordpress-mobile/WordPress-iOS/pull/15798
- [x] Check Small Devices: 🟢 
- [x] Check RTL languages: https://github.com/wordpress-mobile/WordPress-iOS/pull/15810
- [x] Design Review

### PR DESCRIPTION
**Pull request by momo-ozawa**
Part of #15191

- Hooked up restore site APIs
    - Restore site
    - Get rewind status (poll for status **every 5 seconds**)
- Added a new Restore Complete screen

### Merge directions:
- ⚠️ Merge https://github.com/wordpress-mobile/WordPress-iOS/pull/15633 first

### Notes:
- Styling isn't 100% complete
- The `Visit site` button on the Restore Complete screen isn't hooked up yet
- Error handling will be addressed in a separate PR

### To test:

1. My Site > Activity Log > rewindable activity cell
2. Tap on ellipsis icon
3. Tap on `Restore`
4. Tap on `Restore to this point`
5. Tap on `Confirm`
     - ✅ The progress view should update
     - ✅ Once the site has been restored, the app should transition to the Restore Complete screen (Note that it might take a little while to transition after it hits `100%`)
     

30% (`.running`)  | 100% (`.running`) | Complete (`.finished`)
--- | --- | ---
Screen Shot 2021-01-14 at 19 38 26 | Screen Shot 2021-01-14 at 19 38 34 | Screen Shot 2021-01-14 at 19 38 55 

### PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Jetpack/Jetpack Restore/Restore Complete/JetpackRestoreCompleteViewController.swift [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+class JetpackRestoreCompleteViewController: BaseRestoreCompleteViewController {
+
+    // MARK: - Initialization
+
+    override init(site: JetpackSiteRef, activity: Activity) {
+        let restoreCompleteConfiguration = JetpackRestoreCompleteConfiguration(
+            title: NSLocalizedString("Restore", comment: "Title for Jetpack Restore Complete screen"),
+            iconImage: .gridicon(.history),
+            messageTitle: NSLocalizedString("Your site has been restored", comment: "Title for the Jetpack Restore Complete message."),
+            messageDescription: NSLocalizedString("All of your selected itmes are now restored back to %1$@.", comment: "Description for the Jetpack Backup Restore message. %1$@ is a placeholder for the selected date."),
```
Comment (leandroalonso) [inline WordPress/Classes/ViewRelated/Jetpack/Jetpack Restore/Restore Complete/JetpackRestoreCompleteViewController.swift]:
```suggestion
            messageDescription: NSLocalizedString("All of your selected items are now restored back to %1$@.", comment: "Description for the Jetpack Backup Restore message. %1$@ is a placeholder for the selected date."),
```

Typo: itmes

File: WordPress/Classes/Services/JetpackRestoreService.swift [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import Foundation
+
+@objc class JetpackRestoreService: LocalCoreDataService {
+
+    private lazy var serviceV1: ActivityServiceRemote_ApiVersion1_0 = {
+        let api = WordPressComRestApi.defaultApi(in: managedObjectContext,
```
Comment (leandroalonso) [inline WordPress/Classes/Services/JetpackRestoreService.swift]: You can omit the `localeKey`

**Review by leandroalonso — COMMENTED** [review]
@momo-ozawa works as described!

One question though, the confirmation message says: 

Screen Shot 2021-01-14 at 12 56 29

However, this might be not true: what if the user selected just WordPress Themes?

cc @osullivanchris

File: WordPress/Classes/Services/JetpackRestoreService.swift (hunk 2 above)
Comment (momo-ozawa) [reply WordPress/Classes/Services/JetpackRestoreService.swift]: Fixed! 10cd307

File: WordPress/Classes/ViewRelated/Jetpack/Jetpack Restore/Restore Complete/JetpackRestoreCompleteViewController.swift (hunk 1 above)
Comment (momo-ozawa) [reply WordPress/Classes/ViewRelated/Jetpack/Jetpack Restore/Restore Complete/JetpackRestoreCompleteViewController.swift]: Fixed! 24c9a6e

**Comment by momo-ozawa** [comment]
@leandroalonso @osullivanchris 

> However, this might be not true: what if the user selected just WordPress Themes?

Hmm, true. This is what Wordpress.com looks like. 

Screen Shot 2021-01-15 at 10 35 02

**Comment by osullivanchris** [comment]
> However, this might be not true: what if the user selected just WordPress Themes?

Good spot. I pulled the message from something already existing on web. But maybe I didn't grab the right one.

@momo-ozawa in that screenshot, does the message change depending on what you select? 

Either way, my intention was to follow what we are doing on web. So if we have something like this already, that's great, let's use it. If we don't, then we can come up with our own alternative.

**Comment by osullivanchris** [comment]
Discussed this [on Android over here](https://github.com/wordpress-mobile/WordPress-Android/pull/13775#issuecomment-760960333). After discussion with @ashiagr - realised that the word 'All' is really the only one causing issue. 

If we remove the word 'all', the sentence still works grammatically, but it doesn't conclusively say that we are removing *everything* when the user has only selected *some* things. That should resolve the issue I think.

cc @leandroalonso

**Comment by momo-ozawa** [comment]
@leandroalonso (cc: @osullivanchris) 

- Removed the word "all"
- Restore action is disabled if no items are selected

Updated restore warning copy | No items selected | At least 1 item selected 
--- | --- | ---
Simulator Screen Shot - iPhone 12 Pro - 2021-01-16 at 10 39 45 | Simulator Screen Shot - iPhone 12 Pro - 2021-01-16 at 10 39 56 | Simulator Screen Shot - iPhone 12 Pro - 2021-01-16 at 10 40 01

**Review by leandroalonso — APPROVED** [review]

**Pull request merged**